### PR TITLE
Improve documentation and specs in the Function module

### PR DESCRIPTION
**Pull request by whatyouhide**
A bunch of improvements that I was late for making in https://github.com/elixir-lang/elixir/pull/7503 :).

Notable things:

  * I removed the `Function.value/0` type. I think we should be conservative with types since then they're public API and since this type was just defined as `term/0`, I used `term` instead.

  * Added some additional examples.

  * Restructured docs a bit.

  * Fixed the spec for `info/2` (which was defined as `info(fun)`) so that it is defined for `info/2` and so that it encodes the information that the passed item is returned as well.

### CODE REVIEW

File: lib/elixir/lib/function.ex [hunk 1]
```diff
@@ -2,10 +2,10 @@ defmodule Function do
   @moduledoc """
   A set of functions for working with functions.
 
-  We can have two types of captured functions: external and local.
+  There are two types of captured functions: **external** and **local**.
```
Comment (wojtekmach) [inline lib/elixir/lib/function.ex]: s/external/remote?

Comment (josevalim) [reply lib/elixir/lib/function.ex]: This is the function internal naming. So we need to stick with external.

Comment (wojtekmach) [reply lib/elixir/lib/function.ex]: TIL, thanks.

Comment (michalmuskala) [reply lib/elixir/lib/function.ex]: Yes, we should stick with "external" - it's what `:erlang.fun_info(fun, :type)` returns. Although, internally in erts they are called exports.

File: lib/elixir/lib/function.ex [hunk 2]
```diff
@@ -2,10 +2,10 @@ defmodule Function do
...
   External functions are functions residing in modules that are captured
-  with `&/1`, such as `&String.length/1`. Local are the anonymous functions
-  defined with `fn/1` or with the capture operator `&/1` using `&1`, `&2`
+  with `&/1`, such as `&String.length/1`. Local functions are anonymous functions
+  defined with `fn/1` or with the capture operator `&/1` using `&1`, `&2`,
   and so on as replacements.
```
Comment (michalmuskala) [inline lib/elixir/lib/function.ex]: Would it make sense to add an example using `Function.info(fn -> nil end, :type) => :local` and `Function.info(&String.length/1, :type) => :external`?

Comment (whatyouhide) [reply lib/elixir/lib/function.ex]: We have that example in `Function.info/1`, I think it's fine to skip it here.

**Pull request merged**